[Issue 574] Search MVP deliverable spec (#579)
HHS/simpler-grants-gov: documentation/milestones/individual_milestones/search_mvp.md

What functionality do we expect to be in place by ***the end*** of work on this deliverable?

- **[Incident Response](https://github.com/HHS/grants-equity/issues/373):** Ensures that we have a robust incident response plan in place when simpler.grants.gov services are interrupted
- **[Search API](https://github.com/HHS/grants-equity/issues/90):** Enables system-to-system users to access the search features through the API
- **[Search Page](https://github.com/HHS/grants-equity/issues/576):** Enables all other users to access the search features through the user interface
- **[User Research Incentives](https://github.com/HHS/grants-equity/issues/84):** Enables us to engage stakeholders for user testing of the new search functionality

Is there any notable functionality we do ***not*** expect to be in place before works starts on this deliverable?

- **AuthN/AuthZ:** Authentication (AuthN) and Authorization (AuthZ) will be the focus of a subsequent deliverable, so the MVP is not expected to support any features that require users to be logged in.

### Open questions
<!-- Optional -->

#### Does the metadata and the text of the opportunity need to be translated?

> TODO

Reviewer: We will determine if individual fields from the search results can be easily translated in this effort but opportunity listings and all search results will not be translated in this 30k ft deliverable.

> TODO

Reviewer: ```suggestion
We will determine if individual fields from the search results can be easily translated in this effort but opportunity listings and all search results will not be translated in this 30k ft deliverable. 
```

#### What metadata is currently available about each opportunity?

> TODO

Reviewer: ```suggestion
Metadata is being determined in the GET Opportunity 30k deliverable. We plan to use publicly available data only. For example, is_draft state will not be used in this Search feature. 
```

### Not doing
<!-- Optional -->

The following work will *not* be completed as part of this deliverable:

- **Personalization:** Because the MVP will not support a login experience, any personalized features that require authentication (e.g. saving searches) are explicitly descoped from this deliverable.
- **Opportunity Listing Page:** The MVP will not include an opportunity listing page, instead the search results will directly link to the corresponding opportunity listing page on grants.gov. An updated opportunity listing page will be completed in its own 30k ft deliverable.
- **Full Text Search:** The MVP is only expected to include (a subset of) filtering based on opportunity metadata that is already available in the existing grants.gov database, so searching the full text of the opportunity is explicitly descoped from this deliverable.

## Integrations

### Translations
<!-- Required -->

*Does this milestone involve delivering any content that needs translation?*

> TODO

Reviewer: ```suggestion
The translation process will be determined in the 30k ft deliverable in Public Launch: Static Site.
```

*If so, when will English-language content be locked? Then when will translation be started and completed?*

> TODO

Reviewer: ```suggestion
The translations process will be determined in the 30k ft deliverable: Static Site Public Launch.
```

### Services going into PROD for the first time
<!-- Required -->

This can include services going into PROD behind a feature flag that is not turned on.

1. **Search API:** Search functionality will be exposed via the API for the first time in this deliverable
2. **Search UI:** Search functionality will be exposed via the user interface for the first time in this deliverable

### Services being integrated in PROD for the first time
<!-- Required -->

Are there multiple services that are being connected for the first time in PROD?

1. **API + Front-end:** This deliverable represents the first time the simpler.grants.gov front-end and API will be integrated in production

### Data being shared publicly for the first time
<!-- Required -->

Are there any fields being shared publicly that have never been shared in PROD before?

> 1. TODO

Reviewer: ```suggestion
The search data and text will be shared for the first time on the site. 
```

### Security considerations
<!-- Required -->

Does this 30k ft deliverable expose any new attack vectors or expand the attack surface of the product?

> 1. TODO

Reviewer: ```suggestion
We're exposing new endpoints that are all read-only this limits vulnerabilities. 
```

If so, how are we addressing these risks?

> 1. TODO

Reviewer: ```suggestion
We plan to complete a security review and set up any mitigation strategies as needed. 
```

> 1. TODO

Reviewer: ```suggestion
Yes, this effort will share new fields that we have not previously shared before. 
```

  - To receive an incentive if I participate in user research for simpler.grants.gov, so that I am fairly compensated for the time that I invest in testing out new features and answering questions
  - Filters and search terms to be added to the url, so that I can save or share the link to a particular set of search results
- As a **project maintainer**, I want:
  - To be notified when simpler.grants.gov experiences downtime, so that I can quickly troubleshoot the issue with minimal impact on end users
  - The system architecture we create for this deliverable to support future search features as well, so that we don't need to discard and re-build critical infrastructure

## Technical description

### Search API
<!-- Optional -->

The ability to search and filter opportunities should be provided through an API endpoint, which allows the user to:

- Search the list of opportunities with open ended key words
- Filter the search results by *at least* one structured field (e.g. award date, opportunity type, etc.)

> - Sort the search results based on the value of *at least*

Reviewer 2: missing end of this sentence?

Author: Yup, lol! I finished the sentence in [this commit](https://github.com/HHS/simpler-grants-gov/pull/579/commits/0a05c4a125042018476cf6e64e3418b935b1fcee)

Through the design and development of this API endpoint, the team should answer the following questions and record the decisions in one or more ADRs:

- How should search and filter criteria be passed to the API endpoint used for search?
- How should we support basic logic such as greater than, less than, as well as and/or/not conditions?
- Which fields should support filtering and sorting?
- Do we use an existing search service (e.g. Elasticsearch) or build our own?

### Search Page

The primary page we'll need to deliver for the user interface (UI) is a search page, which should allow users to:

- Enter key words to search the list of funding opportunities
- Filter search results by *at least* one structured field (e.g. award date, opportunity type, etc.)
- View a common set of fields for each opportunity returned by the search (e.g. title, award date, summary, etc.)
- Sort the set of results returned by *at least* one structured field (e.g. award date, title, etc.)

### Definition of done
<!-- Required -->

- [ ] Basic requirements:
  - [ ] Code is deployed to `main` & PROD through our CI/CD pipeline

>   - [ ] Services are live in PROD (may be behind feature flag)

Reviewer 2: nit: maybe

Author: Thanks for catching this. I updated it in [this commit](https://github.com/HHS/simpler-grants-gov/pull/579/commits/0a05c4a125042018476cf6e64e3418b935b1fcee)

>   - [ ] Simpler site translations are live in PROD (if necessary). Translations of the search results content coming from grants.gov will not be included in this 30k ft deliverable. 

Reviewer 2: what does "if necessary" mean here?

Author: That might have been hold over language from the template. I took it out from this bullet point.

  - [ ] Data needed for metrics is actively being captured in PROD
  - [ ] Key architectural decisions made about this deliverable are documented publicly
- [ ] Functional requirements:
  - [ ] Users can search for opportunities by key word
  - [ ] Users can filter search results by *at least one* structured field (e.g. award date, opportunity type, etc.)
  - [ ] Users can sort search results by *at least* one structured field (e.g. award date, title, etc.)
  - [ ] All of the search features available in the UI are also available via the API, and vice versa
  - [ ] Search criteria are reflected in the URL so that users can bookmark or share the link to a given set of search results
  - [ ] Users can access the corresponding grants.gov page for an opportunity they find on simpler.grants.gov via link on the search results

### Proposed metrics for measuring goals/value/definition of done
<!-- Required -->

1. Total number of searches made via the UI
2. Total number of searches made via the API
3. Number of results returned per search
4. Number of searches made per search term
5. Number of searches made per filter combination

> 6. Click conversion rate from search results to the opportunity listing (on grants.gov)

Reviewer 2: Eventually it would be great to have metrics about whether the highest-ranked results using our algorithm are the "right" results for users. Like are they clicking on the first result based on their search. 

There might be some good articles or blogs about how search engines like Google evaluate the effectiveness of their search results.

And having supervised learning based on "was this relevant to your search? yes/no".

But those are complex measures, so maybe we defer them to Search 2.0.

Author: Good call, I added a note that we may want to track search result accuracy in the future in [this commit](https://github.com/HHS/simpler-grants-gov/pull/579/commits/0a05c4a125042018476cf6e64e3418b935b1fcee)

# Search MVP

| Field           | Value                                                      |
| --------------- | ---------------------------------------------------------- |
| Document Status | Completed                                                  |
| Epic Link       | [Issue 89](https://github.com/HHS/grants-equity/issues/89) |
| Epic Dashboard  | [Product Roadmap](https://github.com/orgs/HHS/projects/12) |
| Product Owner   | Lucas Brown                                                |
| Document Owner  | Billy Daly                                                 |
| Lead Developer  | Aaron Couch                                                |
| Lead Designer   | Andy Cochran                                               |

## Short description
<!-- Required -->

> - **What:** Launch a minimum viable search feature for Simpler Grants.gov that allows users to filter and sort funding opportunities by *at least* one field and by keywords.

Reviewer: ```suggestion
- **What:** Soft launch a minimum viable search feature for Simpler Grants.gov that allows users to filter and sort funding opportunities by *at least* one field and by keywords.
```

- Enter key words to search the list of funding opportunities
- Filter search results by *at least* one structured field (e.g. award date, opportunity type, etc.)
- View a common set of fields for each opportunity returned by the search (e.g. title, award date, summary, etc.)
- Sort the set of results returned by *at least* one structured field (e.g. award date, title, etc.)

### Opportunity Listing Page

The secondary page we want to deliver within the user interface (UI) is an opportunity listing page, which should:

- Display a consistent list of metadata about each opportunity
- Use only fields that are available in the existing grants.gov cloud database
- Link to the corresponding opportunity in the existing grant.gov interface

**Note:** This is a stretch goal for this deliverable and may be substituted with a direct link to the grants.gov page for a given opportunity from the search results.

### Definition of done
<!-- Required -->

- [ ] Basic requirements:
  - [ ] Code is deployed to `main` & PROD through our CI/CD pipeline
  - [ ] Services are live in PROD (may be behind feature flag)

>   - [ ] Translations are live in PROD (if necessary)

Reviewer: Aaron will be reviewing Monday but not sure about how translations may work cc @acouch

  - [ ] All new services have completed a 508 compliance review (if necessary)
  - [ ] Data needed for metrics is actively being captured in PROD
  - [ ] Key architectural decisions made about this deliverable are documented publicly
- [ ] Functional requirements:
  - [ ] Users can search for opportunities by key word
  - [ ] Users can filter search results by *at least one* structured field (e.g. award date, opportunity type, etc.)
  - [ ] Users can sort search results by *at least* one structured field (e.g. award date, title, etc.)
  - [ ] All of the search features available in the UI are also available via the API, and vice versa
  - [ ] Search criteria are reflected in the URL so that users can bookmark or share the link to a given set of search results
  - [ ] Users can access the corresponding grants.gov page for an opportunity they find on simpler.grants.gov via link on the search results or opportunity listing page

### Proposed metrics for measuring goals/value/definition of done
<!-- Required -->

1. Total number of searches made via the UI
2. Total number of searches made via the API
3. Number of results returned per search

> 4. Frequency distribution of keyword search terms

Reviewer: @widal001 What do you mean by this metric?

Author: We'll want to know what the _most_ and _least_ common search terms are, e.g. are a lot of people searching for "environmental subsidy" or "infrastructure". Having a frequency distribution of searches by keyword would help us map trends in the content and format of searches

Feel free to reword this in a way that makes more sense to y'all

5. Frequency distribution of field-based filters
6. Click conversion rate from search results to opportunity listing page

### Destination for live updating metrics

> <!-- Required -->

Reviewer: What do you think of having a metrics dashboard that's manually updated in the wiki for now, until we have a decision on the public measurement dashboard.

Author: Yeah I like this idea and I think it meets Lucas' need. We have a [similar call out on other pages](https://github.com/HHS/simpler-grants-gov/blob/main/documentation/milestones/individual_milestones/static_site.md#destination-for-live-updating-metrics)

## Integrations

### Translations
<!-- Required -->

*Does this milestone involve delivering any content that needs translation?*

TODO

*If so, when will English-language content be locked? Then when will translation be started and completed?*

TODO

### Services going into PROD for the first time
<!-- Required -->

This can include services going into PROD behind a feature flag that is not turned on.

1. **Search API:** Search functionality will be exposed via the API for the first time in this deliverable
2. **Search UI:** Search functionality will be exposed via the user interface for the first time in this deliverable

### Services being integrated in PROD for the first time
<!-- Required -->

Are there multiple services that are being connected for the first time in PROD?

1. **API + Front-end:** This deliverable represents the first time the simpler.grants.gov front-end and API will be integrated in production

### Data being shared publicly for the first time
<!-- Required -->

Are there any fields being shared publicly that have never been shared in PROD before?

> 1. TODO

Reviewer: The data that is being shared should be the same or similar to what exists on grants.gov. We can determine this in the deliverable and through the security review process.

Author: I think this would be a perfect callout! We might also want to highlight if we anticipate exposing new fields that are currently available on grants.gov cloud but were not released with the `GET /opportunities` endpoint. 

We don't necessarily need to name the fields directly just highlight that there will be net new opportunity fields relative to our previous deliverable (if that's the expectation)

### Security considerations
<!-- Required -->

> Does this milestone expose any new attack vectors or expand the attack surface of the product?

Reviewer: ```suggestion
Does this 30k ft deliverable expose any new attack vectors or expand the attack surface of the product?
```
